### ISSUE TITLE
LibJS: Implement `[[Call]]` closer to the spec (namely `OrdinaryCallBindThis`)

### ISSUE THREAD

**Issue by trflynn89** [SerenityOS/serenity#8509]
Opening an issue for this one for some lingering questions I have.

It seems we don't always have the correct `this` value when invoking functions - sometimes it is undefined when it should be the global `this`. For example, the following will print `undefined` when logging `this`:
```javascript
'a'.replace('a', function() { console.log(this); return 'b'; });
```

[`OrdinaryCallBindThis`](https://tc39.es/ecma262/#sec-ordinarycallbindthis) has a few different paths to set `this` in the callee context depending on the mode (strict, lexical) and provided `this` value when `[[Call]]` is invoked. This diff attempts to implement that spec in VM.cpp:

```diff
diff --git a/Userland/Libraries/LibJS/Runtime/VM.cpp b/Userland/Libraries/LibJS/Runtime/VM.cpp
index fe04633b2..c69142849 100644
--- a/Userland/Libraries/LibJS/Runtime/VM.cpp
+++ b/Userland/Libraries/LibJS/Runtime/VM.cpp
@@ -587,6 +587,33 @@ void VM::prepare_for_ordinary_call(FunctionObject& function, ExecutionContext& c
     // 15. Return calleeContext. (See NOTE above about how contexts are allocated on the C++ stack.)
 }
 
+// 10.2.1.2 OrdinaryCallBindThis ( F, calleeContext, thisArgument ), https://tc39.es/ecma262/#sec-ordinarycallbindthis
+void VM::ordinary_call_bind_this(FunctionObject& function, ExecutionContext& callee_context, Value this_argument)
+{
+    auto this_mode = function.this_mode();
+    auto* callee_realm = function.realm();
+
+    auto* local_environment = callee_context.lexical_environment;
+    auto& function_environment = verify_cast<FunctionEnvironment>(*local_environment);
+
+    if (!callee_realm || this_mode == FunctionObject::ThisMode::Lexical) {
+        function_environment.bind_this_value(function.global_object(), callee_context.this_value);
+        return;
+    }
+
+    Value this_value;
+    if (this_mode == FunctionObject::ThisMode::Strict) {
+        this_value = this_argument;
+    } else if (this_argument.is_nullish()) {
+        auto& global_environment = callee_realm->environment();
+        this_value = global_environment.global_this_value();
+    } else {
+        this_value = this_argument.to_object(function.global_object());
+    }
+
+    function_environment.bind_this_value(function.global_object(), this_value);
+}
+
 Value VM::call_internal(FunctionObject& function, Value this_value, Optional<MarkedValueList> arguments)
 {
     VERIFY(!exception());
@@ -609,11 +636,8 @@ Value VM::call_internal(FunctionObject& function, Value this_value, Optional<Mar
     if (arguments.has_value())
         callee_context.arguments.extend(arguments.value().values());
 
-    if (auto* environment = callee_context.lexical_environment) {
-        auto& function_environment = verify_cast<FunctionEnvironment>(*environment);
-        VERIFY(function_environment.this_binding_status() == FunctionEnvironment::ThisBindingStatus::Uninitialized);
-        function_environment.bind_this_value(function.global_object(), callee_context.this_value);
-    }
+    if (callee_context.lexical_environment)
+        ordinary_call_bind_this(function, callee_context, this_value);
 
     if (exception())
         return {};
```

There's a couple differences in `ordinary_call_bind_this` from the spec though that I'm not quite sure about.
1. There's a null check for `callee_realm`. The spec doesn't indicate this might be null, but it sometimes is.
2. When the mode is lexical (or `callee_realm` is null), it binds `callee_context.this_value` to `this`. This mimics the old behavior in this branch - `test-js` fails left-and-right without this. I don't see where the spec indicates this should happen though, so I'm not sure what do do with that.

We get about 0.2% improvement on test262 with these changes though:
`23127/41267 (56.04%) [ ✅ 23127 ❌ 15324 ⚙️ 124 💀 7 💥️ 2685 ]` (Before)
`23212/41267 (56.25%) [ ✅ 23212 ❌ 15239 ⚙️ 124 💀 8 💥️ 2684 ]` (After)

Thoughts on this implementation / the differences from the spec?

**Comment by davidot** [comment]
Reading the spec I agree that `callee_realm` should never be null (https://tc39.es/ecma262/#sec-code-realms) however this is because the default in `Function::realm()` is `nullptr` so maybe we should make the default be `&global_object()`.

For the second point I'm not sure. But in the spec https://tc39.es/ecma262/#sec-ordinarycallbindthis it does just return NormalCompletion on lexical mode. But looking at the spec I don't see how this should would result in that behavior.

Furthermore this does seem to have two regressions in test262: `test/language/directive-prologue/14.1-14-s.js` and `test/language/directive-prologue/14.1-8-s.js` but that seems to be due to `"use strict";` parsing problems.

**Comment by davidot** [comment]
> Furthermore this does seem to have two regressions in test262: `test/language/directive-prologue/14.1-14-s.js` and `test/language/directive-prologue/14.1-8-s.js` but that seems to be due to `"use strict";` parsing problems.
Actually this is more tricky than I initially assumed but maybe that should be a separate patch.

**Comment by trflynn89** [comment]
Fixed by #8667 :)
